src/ai/backend/manager/models/alembic/versions/d3f8c74bf148_user_main_keypair.py

```diff
@@ -0,0 +1,173 @@
+        )
+        user = sa.Column("user", GUID, sa.ForeignKey("users.uuid"))
+        user_row = relationship("UserRow", back_populates="keypairs", foreign_keys=user)
+
+    class UserRow(Base):  # type: ignore[valid-type, misc]
+        __tablename__ = "users"
+        __table_args__ = {"extend_existing": True}
+
+        uuid = sa.Column("uuid", GUID, primary_key=True)
+        role = sa.Column("role", EnumValueType(UserRole), default=UserRole.USER)
+        email = sa.Column("email", sa.String(length=64))
+        main_access_key = sa.Column(
+            "main_access_key",
+            sa.String(length=20),
+            sa.ForeignKey("keypairs.access_key", ondelete="RESTRICT"),
+            nullable=True,
+        )
+        keypairs = relationship(
+            "KeyPairRow", back_populates="user_row", foreign_keys=KeyPairRow.user
+        )
+        main_keypair = relationship("KeyPairRow", foreign_keys=main_access_key)
+
+    def pick_main_keypair(keypair_list: list[KeyPairRow]) -> KeyPairRow | None:
+        try:
+            return sorted(keypair_list, key=lambda k: k.created_at)[0]
+        except IndexError:
+            return None
+
+    def prepare_keypair(
+        user_email,
+        user_id,
+        user_role,
+    ) -> dict[str, Any]:
+        ak, sk = generate_keypair()
+        pubkey, privkey = generate_ssh_keypair()
+        return {
+            "user_id": user_email,
+            "user": user_id,
+            "access_key": ak,
+            "secret_key": sk,
+            "is_active": True,
+            "is_admin": user_role in (UserRole.SUPERADMIN, UserRole.ADMIN),
+            "resource_policy": DEFAULT_KEYPAIR_RESOURCE_POLICY_NAME,
+            "rate_limit": DEFAULT_KEYPAIR_RATE_LIMIT,
+            "num_queries": 0,
+            "ssh_public_key": pubkey,
+            "ssh_private_key": privkey,
+        }
+
+    connection = op.get_bind()
+    sess_factory = sessionmaker(connection)
+    db_session = sess_factory()
+    while True:
+        user_id_pk_maps = []
+        user_query = (
+            sa.select(UserRow)
+            .where(UserRow.main_access_key.is_(sa.null()))
+            .limit(PAGE_SIZE)
+            .options(selectinload(UserRow.keypairs))
+        )
+        user_rows: list[UserRow] = db_session.scalars(user_query).all()
+
+        if not user_rows:
+            break
+
+        for row in user_rows:
+            primary = pick_main_keypair(row.keypairs)
+            if primary is None:
+                # Create new keypair when the user has no keypair
+                kp_data = prepare_keypair(row.email, row.uuid, row.role)
+                db_session.execute(sa.insert(KeyPairRow).values(**kp_data))
+                user_id_pk_maps.append(
+                    {"user_id": row.uuid, "main_access_key": kp_data["access_key"]}
+                )
+            else:
+                user_id_pk_maps.append({"user_id": row.uuid, "main_access_key": primary.access_key})
```

Comment:
Just a very minor fix, but I think the variable `primary` should be `main_keypair` or `main_kp`. 😉